addons/project/controllers/project_sharing_chatter.py

```diff
@@ -0,0 +1,64 @@
+from odoo.tools import consteq
+
+from odoo.addons.portal.controllers.mail import PortalChatter
+from .portal import ProjectCustomerPortal
+
+
+class ProjectSharingChatter(PortalChatter):
+    def _check_project_sharing_access(self, project_id, res_model, res_id, token):
+        """ Check if the chatter in project sharing can be accessed
+
+            If the portal user is in the project sharing, then we do not have the access token of the task
+            but we can have the one of the project (if the user accessed to the project sharing views via the shared link).
+            So, we need to check if the chatter is for a task and if the res_id is a task
+            in the project shared. Then, if we had the project token and this one is the one in the project
+            then we return the token of the task to continue the portal chatter process.
+            If we do not have any token, then we need to check if the portal user is a follower of the project shared.
+            If it is the case, then we give the access token of the task.
+        """
+        project_sudo = ProjectCustomerPortal._document_check_access(self, 'project.project', project_id, token)
+        can_access = project_sudo and res_model == 'project.task' and res_id in project_sudo.tasks.ids and project_sudo._check_project_sharing_access('comment')
+        if token:
+            can_access &= consteq(project_sudo.access_token, token)
+        else:
+            can_access &= request.env.user.partner_id.commercial_partner_id in project_sudo.message_partner_ids
+        if not can_access:
+            raise Forbidden()
+        task = project_sudo.tasks.filtered(lambda task: task.id == res_id)
+        return task[task._mail_post_token_field]
+
+    @route()
+    def portal_chatter_init(self, res_model, res_id, domain=False, limit=False, **kwargs):
+        project_sharing_id = kwargs.get('project_sharing_id')
+        if project_sharing_id:
+            # if there is a token in `kwargs` then it should be the access_token of the project shared
+            token = self._check_project_sharing_access(project_sharing_id, res_model, res_id, kwargs.get('token'))
+            if token:
+                del kwargs['project_sharing_id']
+                kwargs['token'] = token
+        return super().portal_chatter_init(res_model, res_id, domain=domain, limit=limit, **kwargs)
+
+    @route()
+    def portal_chatter_post(self, res_model, res_id, message, attachment_ids=None, attachment_tokens=None, **kw):
+        project_sharing_id = kw.get('project_sharing_id')
+        if project_sharing_id:
+            token = self._check_project_sharing_access(project_sharing_id, res_model, res_id, kw.get('token'))
+            if token:
+                del kw['project_sharing_id']
+                kw['token'] = token
+        return super().portal_chatter_post(res_model, res_id, message, attachment_ids=attachment_ids, attachment_tokens=attachment_tokens, **kw)
```

Comment:
considering that portal users can bypass this entirely and directly call `message_post()`, it would be interesting to evaluate the riskss/differences between the 2, and make a comment here to explain why this is considered safe.
possibly explain that in the super `portal_chatter_post()` code too.